File: Editorial/0030_Substring_with_Concatenation_of_All_Words/CSharp/1_Check_All_Indices_Using_a_Hash_Table.cs
public class Solution {
    private Dictionary<string, int> wordCount = new Dictionary<string, int>();
    private int wordLength;
    private int substringSize;
    private int k;

    private bool Check(int i, string s) {
        // Copy the original dictionary to use for this index
        var remaining = new Dictionary<string, int>(wordCount);
        int wordsUsed = 0;
        // Each iteration will check for a match in words
        for (int j = i; j < i + substringSize; j += wordLength) {
            string sub = s.Substring(j, wordLength);
            if (remaining.ContainsKey(sub) && remaining[sub] != 0) {
                remaining[sub] = remaining[sub] - 1;
                wordsUsed++;
            } else {
                break;
            }
        }

        return wordsUsed == k;
    }

    public IList<int> FindSubstring(string s, string[] words) {
        int n = s.Length;
        k = words.Length;
        wordLength = words[0].Length;
        substringSize = wordLength * k;
        foreach (var word in words)
            wordCount[word] =
                wordCount.ContainsKey(word) ? wordCount[word] + 1 : 1;
        IList<int> answer = new List<int>();
        for (int i = 0; i < n - substringSize + 1; i++) {
            if (Check(i, s)) {
                answer.Add(i);
            }
        }

        return answer;
    }
}

File: Editorial/30_Substring_with_Concatenation_of_All_Words/TypeScript/1_Check_All_Indices_Using_a_Hash_Table.ts
function findSubstring(s: string, words: string[]): number[] {
    const n = s.length;
    const k = words.length;
    const wordLength = words[0].length;
    const substringSize = wordLength * k;
    const wordCount: Record<string, number> = {};
    words.forEach((word) => {
        wordCount[word] = (wordCount[word] || 0) + 1;
    });
    const check = (i: number): boolean => {
        const remaining: Record<string, number> = { ...wordCount };
        let wordsUsed = 0;
        for (let j = i; j < i + substringSize; j += wordLength) {
            const sub = s.slice(j, j + wordLength);
            if (remaining[sub]) {
                remaining[sub]--;
                wordsUsed++;
            } else {
                break;
            }
        }
        return wordsUsed === k;
    };
    const answer: number[] = [];
    for (let i = 0; i < n - substringSize + 1; i++) {
        if (check(i)) {
            answer.push(i);
        }
    }
    return answer;
}
